mods/default/nodes.lua

```diff
@@ -1986,6 +1986,45 @@ minetest.register_node("default:sand_with_kelp", {
 -- Corals
 --
 
+local function coral_on_place(itemstack, placer, pointed_thing)
+	if pointed_thing.type ~= "node" or not placer then
+		return itemstack
+	end
+
+	local player_name = placer:get_player_name()
+	local pos_under = pointed_thing.under
+	local pos_above = pointed_thing.above
+	local node_under = minetest.get_node(pos_under)
+	local def_under = minetest.registered_nodes[node_under.name]
+
+	if def_under and def_under.on_rightclick then
+		return def_under.on_rightclick(pos_under, node_under.name,
+				placer, itemstack, pointed_thing) or itemstack
+	end
+
+	if node_under.name ~= "default:coral_skeleton" or
+			minetest.get_node(pos_above).name ~= "default:water_source" then
+		return itemstack
+	end
+
+	if minetest.is_protected(pos_under, player_name) or
+			minetest.is_protected(pos_above, player_name) then
+		minetest.log("action", player_name
+			.. " tried to place " .. itemstack:get_name()
+			.. " at protected position "
+			.. minetest.pos_to_string(pos_under))
+		minetest.record_protection_violation(pos_under, player_name)
+		return itemstack
+	end
+
+	minetest.swap_node(pos_under, {name = itemstack:get_name()})
```

Comment:
I'm not sure we should use 'swap node', as coral skeleton might have metadata added to it for some reason.

Reply:
If mods inject arbitrary metadata into nodes it's their job to clean up their garbage after swapping. Swap node was made for situations like this, where metadata is either not present or has to be contained in the destination node.
You probably argued the same way with the moss growth, which would be far more important to use swap_node. Default does not need to spend more time on cleaning up garbage of other mods.

Reply:
Thanks.